electron/src/js/lib/SingleSignOn.ts

```diff
@@ -0,0 +1,318 @@
+
+class SingleSignOn {
+  private static readonly ALLOWED_BACKEND_ORIGINS: string[] = [
+    'https://staging-nginz-https.zinfra.io',
+    'https://prod-nginz-https.wire.com',
+  ];
+  private static readonly PRELOAD_SSO_JS = path.join(app.getAppPath(), 'dist', 'js', 'preload-sso.js');
+  private static readonly SINGLE_SIGN_ON_FRAME_NAME = 'WIRE_SSO';
+  private static readonly SSO_PROTOCOL = 'wire-sso';
+  private static readonly SSO_PROTOCOL_HOST = 'response';
+  private static readonly SSO_PROTOCOL_RESPONSE_SIZE_LIMIT = 255;
+  private static readonly SSO_SESSION_NAME = 'sso';
+
+  private static readonly RESPONSE_TYPES = {
+    AUTH_ERROR_COOKIE: 'AUTH_ERROR_COOKIE',
+    AUTH_ERROR_SESS_NOT_AVAILABLE: 'AUTH_ERROR_SESS_NOT_AVAILABLE',
+    AUTH_SUCCESS: 'AUTH_SUCCESS',
+  };
+
+  public static loginAuthorizationSecret: string | undefined;
+
+  private session: Electron.Session | undefined;
+  private readonly mainSession: Electron.Session;
+  private readonly senderWebContents: Electron.WebContents;
+  private readonly windowOriginUrl: URL;
+
+  private static readonly cookies = {
+    copy: async (from: Electron.Session, to: Electron.Session, url: URL) => {
+      const cookies = await SingleSignOn.cookies.getBackendCookies(from, url);
+      for (const cookie of cookies) {
+        await SingleSignOn.cookies.setCookie(to, cookie, url.toString());
+      }
+      await SingleSignOn.cookies.flushCookies(to);
+    },
+    flushCookies: (session: Electron.Session): Promise<void> => {
+      return new Promise((resolve, reject) => {
+        session.cookies.flushStore(() => resolve());
+      });
+    },
+    getBackendCookies: (session: Electron.Session, url: URL): Promise<Electron.Cookie[]> => {
+      return new Promise((resolve, reject) => {
+        const rootDomain = url.hostname
+          .split('.')
+          .reverse()
+          .splice(0, 2)
+          .reverse()
+          .join('.');
+        session.cookies.get({domain: rootDomain, secure: true}, (error, cookies) => {
+          if (error) {
+            return reject(error);
+          }
+          resolve(cookies);
+        });
+      });
+    },
+    setCookie: (session: Electron.Session, cookie: Electron.Cookie, url: string) => {
+      return new Promise((resolve, reject) => {
+        session.cookies.set({url, ...(<Electron.Details>cookie)}, error => {
+          if (error) {
+            return reject(error);
+          }
+          resolve();
+        });
+      });
+    },
+  };
+
+  private static readonly protocol = {
+    generateSecret: (length: number): Promise<string> => {
+      return new Promise((resolve, reject) => {
+        crypto.randomBytes(length, (error, bytes) => {
+          if (error) {
+            return reject(error);
+          }
+
+          resolve(bytes.toString('hex'));
+        });
+      });
+    },
+    register: async (session: Electron.Session, callback: (type: string) => void): Promise<void> => {
```

Comment:
`async` and callback? Can't we return the `type` in the promise?

Reply:
We only wait the function to register the protocol, not to return the type. The callback may be called multiple times.